_posts/2017-04-27-Rails-5-1-final.markdown

```diff
@@ -0,0 +1,110 @@
+---
+layout: post
+title: 'Rails 5.1: Loving JavaScript, System Tests, Encrypted Secrets, and more'
+category: releases
+author: dhh
+published: true
+date: 2017-04-27 09:00:00 -09:00
+---
+In celebration of the 12th [RailsConf](http://railsconf.com) in Phoenix, Arizona this week, we're proud to announce that Rails 5.1 is ready in its final form! We've spent [over 4,100 commits since Rails 5.0](https://github.com/rails/rails/compare/v5.0.0...5-1-stable/) making everything EASIER, SIMPLER, and, uhhh, FUNNER? (That's a RailsConf joke).
+
+The highlight reel hasn't really changed since the first beta, but here's a repeat:
+
+
+## Loving JavaScript
+
+We've had a stormy, perhaps even contentious, relationship with JavaScript over the years. But that time is past. JavaScript has improved immensely over the past few years, particularly with the advent of ES6, and with package and compilation tools like Yarn and Webpack. Rails is embracing both of these solutions with open arms and letting whatever past water flow under the bridge.
+
+JavaScript and Ruby share a deep philosophical bond over language design, if not ecosystem management. Let's focus on the aspects we have in common and help Rails programmers extract the best from JavaScript with the help of some key guiding conventions.
+
+The improvements in Rails 5.1 focus on three major parts:
+
+1. **Manage JavaScript dependencies from NPM via Yarn.** Think of [Yarn](https://yarnpkg.com/en/) like Bundler for JavaScript (it even has [Yehuda Katz](https://github.com/wycats) involved!). This makes it easy to depend on libraries like React or anything else from NPM. Everything you depend on via Yarn is then made available to be required in the asset pipeline, just like vendored dependencies would have been. Just use the binstub `bin/yarn` to add dependencies.
+
+2. **Optionally compile JavaScript with Webpack.** While there are a million different module bundlers/compilers for JavaScript, [Webpack](https://webpack.github.io/) is quickly emerging as the preeminent choice. We've made it easy to use Webpack with Rails through the new [Webpacker](https://github.com/rails/webpacker) gem that you can configure automatically on new projects with `--webpack` (or even `--webpack=react`, `--webpack=angular`, or `--webpack=vue` for a tailored configuration). This is fully compatible with the asset pipeline, which you can continue to use for images, fonts, sounds, whatever. You can even have some JavaScript on the asset pipeline and some done via Webpack. It's all managed via Yarn that's on by default.
+
+3. **Drop jQuery as a default dependency.** We used to require jQuery in order to provide features like data-remote, data-confirm, and the other parts of Rails UJS. This dependency is no longer necessary as we've [rewritten rails-ujs to use vanilla JavaScript](https://github.com/rails/rails-ujs). You're of course still free to use jQuery, but you no longer have to.
```

Comment:
Think we'll eventually need a better location than GitHub for our JS docs. — @fxn should we just turn our JS docs into guides?

I'll update the link, since it's more correct than the rails/rails-ujs repo after the inlining.